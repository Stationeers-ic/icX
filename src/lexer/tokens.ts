// identifier, keyword, delimiter, operator, literal, and comment.
export const TOKEN_TYPES = {
	NEWLINE: "NEWLINE",
	LEFT_BRACE: "LEFT_BRACE",
	RIGHT_BRACE: "RIGHT_BRACE",
	LEFT_PARENTHESIS: "LEFT_PARENTHESIS",
	RIGHT_PARENTHESIS: "RIGHT_PARENTHESIS",
	COLON: "COLON",
	SEMICOLON: "SEMICOLON",
	DOT: "DOT",
	COMMA: "COMMA",
	INCREMENT: "INCREMENT",
	DECREMENT: "DECREMENT",
	ARROW: "ARROW",
	// Relational operators
	LESS_THAN: "LESS_THAN",
	GREATER_THAN: "GREATER_THAN",
	// Equality operators
	EQUAL: "EQUAL",
	NOT_EQUAL: "NOT_EQUAL",
	//assign
	ASSIGNMENT: "ASSIGNMENT",
	EXPONENTIATION_ASSIGNMENT: "EXPONENTIATION_ASSIGNMENT",
	MULTIPLICATION_ASSIGNMENT: "MULTIPLICATION_ASSIGNMENT",
	DIVISION_ASSIGNMENT: "DIVISION_ASSIGNMENT",
	REMAINDER_ASSIGNMENT: "REMAINDER_ASSIGNMENT",
	ADDITION_ASSIGNMENT: "ADDITION_ASSIGNMENT",
	SUBTRACTION_ASSIGNMENT: "SUBTRACTION_ASSIGNMENT",
	//bitwise assignment
	BITWISE_AND_ASSIGNMENT: "BITWISE_AND_ASSIGNMENT",
	BITWISE_OR_ASSIGNMENT: "BITWISE_OR_ASSIGNMENT",
	BITWISE_XOR_ASSIGNMENT: "BITWISE_XOR_ASSIGNMENT",
	BITWISE_LEFT_SHIFT_ASSIGNMENT: "BITWISE_LEFT_SHIFT_ASSIGNMENT",
	BITWISE_RIGHT_SHIFT_ARITHMETIC_ASSIGNMENT: "BITWISE_RIGHT_SHIFT_ARITHMETIC_ASSIGNMENT",
	BITWISE_RIGHT_SHIFT_LOGICAL_ASSIGNMENT: "BITWISE_RIGHT_SHIFT_LOGICAL_ASSIGNMENT",
	// logical assignment
	LOGICAL_AND_ASSIGNMENT: "LOGICAL_AND_ASSIGNMENT",
	LOGICAL_OR_ASSIGNMENT: "LOGICAL_OR_ASSIGNMENT",
	// Arithmetic operators
	EXPONENTIATION: "EXPONENTIATION",
	MULTIPLICATION: "MULTIPLICATION",
	DIVISION: "DIVISION",
	REMAINDER: "REMAINDER",
	ADDITION: "ADDITION",
	SUBTRACTION: "SUBTRACTION",
	// Logical operators
	LOGICAL_AND: "LOGICAL_AND",
	LOGICAL_OR: "LOGICAL_OR",
	LOGICAL_NOT: "LOGICAL_NOT",
	// Bitwise operators
	BITWISE_AND: "BITWISE_AND",
	BITWISE_OR: "BITWISE_OR",
	BITWISE_XOR: "BITWISE_XOR",
	BITWISE_NOT: "BITWISE_NOT",
	BITWISE_RIGHT_SHIFT_LOGICAL: "BITWISE_RIGHT_SHIFT_LOGICAL",
	BITWISE_LEFT_SHIFT: "BITWISE_LEFT_SHIFT",
	BITWISE_RIGHT_SHIFT_ARITHMETIC: "BITWISE_RIGHT_SHIFT_ARITHMETIC",

	LESS_THAN_EQUAL: "LESS_THAN_EQUAL",
	GREATER_THAN_EQUAL: "GREATER_THAN_EQUAL",
	// Keywords
	IF: "IF",
	ELSE: "ELSE",
	SWITCH: "SWITCH",
	CASE: "CASE",
	DEFAULT: "DEFAULT",
	DO: "DO",
	WHILE: "WHILE",
	FOR: "FOR",
	BREAK: "BREAK",
	CONTINUE: "CONTINUE",
	LET: "LET",
	CONST: "CONST",
	FUNCTION: "FUNCTION",
	RETURN: "RETURN",
	TRUE: "TRUE",
	FALSE: "FALSE",
	NULL: "NULL",
	STRING: "STRING",
	NUMBER: "NUMBER",
	IDENTIFIER: "IDENTIFIER",
	COMMENT: "COMMENT",
	// Special
	UNKNOWN: "UNKNOWN",
	EOF: "EOF",
	ERROR: "ERROR",
} as const
export type TOKEN_TYPES = (typeof TOKEN_TYPES)[keyof typeof TOKEN_TYPES]

export const TOKENS: Array<
	| {
			token: TOKEN_TYPES
			patternType: "string"
			pattern: string
	  }
	| {
			token: TOKEN_TYPES
			patternType: "range"
			open: string
			close: string
			ignore?: string[]
			errorOn?: string[]
	  }
	| {
			token: TOKEN_TYPES
			patternType: "function"
			pattern: (position: number, total: string) => [length: number, value: any, error?: string] | null
	  }
> = [
	{
		token: TOKEN_TYPES.EOF,
		patternType: "string",
		pattern: "\u0003",
	},
	{
		token: TOKEN_TYPES.NEWLINE,
		patternType: "string",
		pattern: "\n",
	},
	{
		token: TOKEN_TYPES.LEFT_BRACE,
		patternType: "string",
		pattern: "{",
	},
	{
		token: TOKEN_TYPES.RIGHT_BRACE,
		patternType: "string",
		pattern: "}",
	},
	{
		token: TOKEN_TYPES.LEFT_PARENTHESIS,
		patternType: "string",
		pattern: "(",
	},
	{
		token: TOKEN_TYPES.RIGHT_PARENTHESIS,
		patternType: "string",
		pattern: ")",
	},
	{
		token: TOKEN_TYPES.COLON,
		patternType: "string",
		pattern: ":",
	},
	{
		token: TOKEN_TYPES.SEMICOLON,
		patternType: "string",
		pattern: ";",
	},
	{
		token: TOKEN_TYPES.DOT,
		patternType: "string",
		pattern: ".",
	},
	{
		token: TOKEN_TYPES.COMMA,
		patternType: "string",
		pattern: ",",
	},
	{
		token: TOKEN_TYPES.INCREMENT,
		patternType: "string",
		pattern: "++",
	},
	{
		token: TOKEN_TYPES.DECREMENT,
		patternType: "string",
		pattern: "--",
	},
	{
		token: TOKEN_TYPES.ARROW,
		patternType: "string",
		pattern: "=>",
	},
	{
		token: TOKEN_TYPES.COMMENT,
		patternType: "range",
		open: "/*",
		close: "*/",
	},
	{
		token: TOKEN_TYPES.COMMENT,
		patternType: "function",
		pattern: (position, total) => {
			const reg = /^\/\/[^\n\x03]*/
			const match = reg.exec(total.slice(position))
			// if start position is 0 return 0
			// else return length of match
			if (match === null) return null
			return [match[0].length, match[0].slice(2)]
		},
	},
	{
		token: TOKEN_TYPES.LESS_THAN_EQUAL,
		patternType: "string",
		pattern: "<=",
	},
	{
		token: TOKEN_TYPES.GREATER_THAN_EQUAL,
		patternType: "string",
		pattern: ">=",
	},
	{
		token: TOKEN_TYPES.EQUAL,
		patternType: "string",
		pattern: "==",
	},
	{
		token: TOKEN_TYPES.NOT_EQUAL,
		patternType: "string",
		pattern: "!=",
	},
	{
		token: TOKEN_TYPES.ASSIGNMENT,
		patternType: "string",
		pattern: "=",
	},
	{
		token: TOKEN_TYPES.EXPONENTIATION_ASSIGNMENT,
		patternType: "string",
		pattern: "**=",
	},
	{
		token: TOKEN_TYPES.MULTIPLICATION_ASSIGNMENT,
		patternType: "string",
		pattern: "*=",
	},
	{
		token: TOKEN_TYPES.DIVISION_ASSIGNMENT,
		patternType: "string",
		pattern: "/=",
	},
	{
		token: TOKEN_TYPES.REMAINDER_ASSIGNMENT,
		patternType: "string",
		pattern: "%=",
	},
	{
		token: TOKEN_TYPES.ADDITION_ASSIGNMENT,
		patternType: "string",
		pattern: "+=",
	},
	{
		token: TOKEN_TYPES.SUBTRACTION_ASSIGNMENT,
		patternType: "string",
		pattern: "-=",
	},
	{
		token: TOKEN_TYPES.BITWISE_AND_ASSIGNMENT,
		patternType: "string",
		pattern: "&=",
	},
	{
		token: TOKEN_TYPES.BITWISE_OR_ASSIGNMENT,
		patternType: "string",
		pattern: "|=",
	},
	{
		token: TOKEN_TYPES.BITWISE_XOR_ASSIGNMENT,
		patternType: "string",
		pattern: "^=",
	},
	{
		token: TOKEN_TYPES.BITWISE_LEFT_SHIFT_ASSIGNMENT,
		patternType: "string",
		pattern: "<<=",
	},
	{
		token: TOKEN_TYPES.BITWISE_RIGHT_SHIFT_ARITHMETIC_ASSIGNMENT,
		patternType: "string",
		pattern: ">>=",
	},
	{
		token: TOKEN_TYPES.BITWISE_RIGHT_SHIFT_LOGICAL_ASSIGNMENT,
		patternType: "string",
		pattern: ">>>=",
	},
	{
		token: TOKEN_TYPES.LOGICAL_AND_ASSIGNMENT,
		patternType: "string",
		pattern: "&&=",
	},
	{
		token: TOKEN_TYPES.LOGICAL_OR_ASSIGNMENT,
		patternType: "string",
		pattern: "||=",
	},
	{
		token: TOKEN_TYPES.EXPONENTIATION,
		patternType: "string",
		pattern: "**",
	},
	{
		token: TOKEN_TYPES.MULTIPLICATION,
		patternType: "string",
		pattern: "*",
	},
	{
		token: TOKEN_TYPES.DIVISION,
		patternType: "string",
		pattern: "/",
	},
	{
		token: TOKEN_TYPES.REMAINDER,
		patternType: "string",
		pattern: "%",
	},
	{
		token: TOKEN_TYPES.ADDITION,
		patternType: "string",
		pattern: "+",
	},
	{
		token: TOKEN_TYPES.SUBTRACTION,
		patternType: "string",
		pattern: "-",
	},
	{
		token: TOKEN_TYPES.LOGICAL_AND,
		patternType: "string",
		pattern: "&&",
	},
	{
		token: TOKEN_TYPES.LOGICAL_OR,
		patternType: "string",
		pattern: "||",
	},
	{
		token: TOKEN_TYPES.LOGICAL_NOT,
		patternType: "string",
		pattern: "!",
	},
	{
		token: TOKEN_TYPES.BITWISE_AND,
		patternType: "string",
		pattern: "&",
	},
	{
		token: TOKEN_TYPES.BITWISE_OR,
		patternType: "string",
		pattern: "|",
	},
	{
		token: TOKEN_TYPES.BITWISE_XOR,
		patternType: "string",
		pattern: "^",
	},
	{
		token: TOKEN_TYPES.BITWISE_NOT,
		patternType: "string",
		pattern: "~",
	},
	{
		token: TOKEN_TYPES.BITWISE_LEFT_SHIFT,
		patternType: "string",
		pattern: "<<",
	},
	{
		token: TOKEN_TYPES.BITWISE_RIGHT_SHIFT_LOGICAL,
		patternType: "string",
		pattern: ">>>",
	},
	{
		token: TOKEN_TYPES.BITWISE_RIGHT_SHIFT_ARITHMETIC,
		patternType: "string",
		pattern: ">>",
	},
	{
		token: TOKEN_TYPES.LESS_THAN,
		patternType: "string",
		pattern: "<",
	},
	{
		token: TOKEN_TYPES.GREATER_THAN,
		patternType: "string",
		pattern: ">",
	},
	{
		token: TOKEN_TYPES.IF,
		patternType: "string",
		pattern: "if",
	},
	{
		token: TOKEN_TYPES.ELSE,
		patternType: "string",
		pattern: "else",
	},
	{
		token: TOKEN_TYPES.SWITCH,
		patternType: "string",
		pattern: "switch",
	},
	{
		token: TOKEN_TYPES.CASE,
		patternType: "string",
		pattern: "case",
	},
	{
		token: TOKEN_TYPES.DEFAULT,
		patternType: "string",
		pattern: "default",
	},
	{
		token: TOKEN_TYPES.DO,
		patternType: "string",
		pattern: "do",
	},
	{
		token: TOKEN_TYPES.WHILE,
		patternType: "string",
		pattern: "while",
	},
	{
		token: TOKEN_TYPES.FOR,
		patternType: "string",
		pattern: "for",
	},
	{
		token: TOKEN_TYPES.BREAK,
		patternType: "string",
		pattern: "break",
	},
	{
		token: TOKEN_TYPES.CONTINUE,
		patternType: "string",
		pattern: "continue",
	},
	{
		token: TOKEN_TYPES.LET,
		patternType: "string",
		pattern: "var",
	},
	{
		token: TOKEN_TYPES.LET,
		patternType: "string",
		pattern: "let",
	},
	{
		token: TOKEN_TYPES.CONST,
		patternType: "string",
		pattern: "const",
	},
	{
		token: TOKEN_TYPES.FUNCTION,
		patternType: "string",
		pattern: "function",
	},
	{
		token: TOKEN_TYPES.RETURN,
		patternType: "string",
		pattern: "return",
	},
	{
		token: TOKEN_TYPES.TRUE,
		patternType: "string",
		pattern: "true",
	},
	{
		token: TOKEN_TYPES.FALSE,
		patternType: "string",
		pattern: "false",
	},
	{
		token: TOKEN_TYPES.NULL,
		patternType: "string",
		pattern: "null",
	},
	{
		token: TOKEN_TYPES.STRING,
		patternType: "range",
		open: '"',
		close: '"',
		ignore: ['\\"'],
		errorOn: ["\n"],
	},
	{
		token: TOKEN_TYPES.STRING,
		patternType: "range",
		open: "'",
		close: "'",
		ignore: ["\\'"],
		errorOn: ["\n"],
	},
	{
		token: TOKEN_TYPES.NUMBER,
		patternType: "function",
		// pattern: /-?\d+(?:\.\d+)?/,
		pattern: (position, total) => {
			const reg = /^\d[_\d]*(?:\.[_\d]+)?/
			const match = reg.exec(total.slice(position))
			// if start position is 0 return 0
			// else return length of match
			if (match === null) return null
			if (isNaN(Number(match[0].replaceAll("_", "")))) return null
			return [match[0].length, Number(match[0].replaceAll("_", ""))]
		},
	},
	{
		token: TOKEN_TYPES.IDENTIFIER,
		patternType: "function",
		// pattern: /[a-zA-Z_$][a-zA-Z0-9_$]*/,
		pattern: (position, total) => {
			const reg = /^[a-zA-Z_]\w*/
			const match = reg.exec(total.slice(position))
			// if start position is 0 return 0
			// else return length of match
			if (match === null) return null
			return [match[0].length, match[0]]
		},
	},
]
const arr: string[] = ["ERROR", "UNKNOWN"]
for (const i in TOKENS) {
	arr.push(TOKENS[i].token)
}
for (const val of Object.values(TOKEN_TYPES)) {
	if (!arr.includes(val)) console.error(val)
}
