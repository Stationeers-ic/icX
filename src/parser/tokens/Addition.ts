import TOKEN_TYPES, { Token, TokenInterface } from "../tokens"
import { getNextTokenFromMathTree, mathTree } from "../evaluateMath"
import { MathToken } from "./mathTokens"


export class Addition extends MathToken implements TokenInterface {
	readonly type = TOKEN_TYPES.Addition
	constructor(left: mathTree, right: mathTree, parent: TokenInterface) {
		super(left, right, parent)
	}
	static is(token: TokenInterface | null): token is Addition {
		if (token === null) return false
		return token.type === TOKEN_TYPES.Addition
	}
}

export default Addition
