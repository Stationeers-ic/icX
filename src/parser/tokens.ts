import { type Token as _LexerToken, TOKEN_TYPES as _LexerTOKEN_TYPES } from "../lexer"
import { ErrorListing } from "./errors"
import { mathTree } from "./evaluateMath"
import { HalfMathToken } from './tokens/math/HalfMathToken';

export type ExtendedLexerToken =
	| _LexerToken<"FUNCTION_CALL", LexerToken[]>
	| _LexerToken<"UNARY_ADDITION">
	| _LexerToken<"UNARY_NEGATION">
export type LexerToken = _LexerToken | ExtendedLexerToken
export const LexerTOKEN_TYPES = {
	..._LexerTOKEN_TYPES,
	FUNCTION_CALL: "FUNCTION_CALL",
	UNARY_ADDITION: "UNARY_ADDITION",
	UNARY_NEGATION: "UNARY_NEGATION",
} as const
export type LexerTOKEN_TYPES = (typeof LexerTOKEN_TYPES)[keyof typeof LexerTOKEN_TYPES]
export function isLexerToken(x?: LexerToken | null): x is _LexerToken {
	if (!x) return false
	return x?.type in _LexerTOKEN_TYPES
}
export function isLexerTOKEN_TYPES(x?: LexerTOKEN_TYPES | null): x is _LexerTOKEN_TYPES {
	if (!x) return false
	return x in _LexerTOKEN_TYPES
}
export interface TokenInterface {
	type: TOKEN_TYPES
	start: number
	end: number
	length: number
	parent: TokenInterface | null
	errors: ErrorListing[]
	isCodeBlock: boolean
	variables?: {}
	constants?: {}
}
export abstract class Token {
	static parse(tokens: LexerToken[], parent: TokenInterface): [TokenInterface, LexerToken[]] | null {
		throw new Error("Method not used in this class")
	}
}

export const HalfMathTokenTOKEN_TYPES = {
	LogicalNot: "LogicalNot",
	BitwiseNot: "BitwiseNot",
	UnaryPlus: "UnaryPlus",
	UnaryMinus: "UnaryMinus",
} as const
export type HalfMathTokenTOKEN_TYPES = (typeof HalfMathTokenTOKEN_TYPES)[keyof typeof HalfMathTokenTOKEN_TYPES]
export const MathOperatorTOKEN_TYPES = {
	Exponentiation: "Exponentiation",
	Multiplication: "Multiplication",
	Dot: "Dot",
	Division: "Division",
	Remainder: "Remainder",
	Addition: "Addition",
	Subtraction: "Subtraction",
	LessThan: "LessThan",
	GreaterThan: "GreaterThan",
	Equal: "Equal",
	NotEqual: "NotEqual",
	LessThanEqual: "LessThanEqual",
	GreaterThanEqual: "GreaterThanEqual",
	LogicalAnd: "LogicalAnd",
	LogicalOr: "LogicalOr",
	BitwiseAnd: "BitwiseAnd",
	BitwiseOr: "BitwiseOr",
	BitwiseXor: "BitwiseXor",
	LeftShift: "LeftShift",
	RightShiftArithmetic: "RightShiftArithmetic",
	RightShiftLogical: "RightShiftLogical",
} as const
export type MathOperatorTOKEN_TYPES = (typeof MathOperatorTOKEN_TYPES)[keyof typeof MathOperatorTOKEN_TYPES]
export const GeneralTOKEN_TYPES = {
	Program: "Program",
	CodeBlock: "CodeBlock",
	VariableDeclaration: "VariableDeclaration",
	VariableDeclarator: "VariableDeclarator",
	VariableAssignment: "VariableAssignment",
	ConstantDeclaration: "ConstantDeclaration",
	ConstantDeclarator: "ConstantDeclarator",
	ConstantAssignment: "ConstantAssignment",
	AssignmentBlock: "AssignmentBlock",
	Identifier: "Identifier",
	StringToken: "StringToken",
	NumberToken: "NumberToken",
	FunctionCall: "FunctionCall",
	FunctionArgument: "FunctionArgument",
	FunctionStatement: "FunctionStatement",
} as const
export type GeneralTOKEN_TYPES = (typeof GeneralTOKEN_TYPES)[keyof typeof GeneralTOKEN_TYPES]

export const TOKEN_TYPES = {
	...MathOperatorTOKEN_TYPES,
	...HalfMathTokenTOKEN_TYPES,
	...GeneralTOKEN_TYPES,
} as const
export type TOKEN_TYPES = (typeof TOKEN_TYPES)[keyof typeof TOKEN_TYPES]
export default TOKEN_TYPES

export const lexerBrakeTokens = {
	NEWLINE: LexerTOKEN_TYPES.NEWLINE,
	SEMICOLON: LexerTOKEN_TYPES.SEMICOLON,
	EOF: LexerTOKEN_TYPES.EOF,
} as const
export const lexerSeparatorTokens = {
	COMMA: LexerTOKEN_TYPES.COMMA,
} as const

export const lexerCalculationTokens = {
	LEFT_PARENTHESIS: LexerTOKEN_TYPES.LEFT_PARENTHESIS,
	RIGHT_PARENTHESIS: LexerTOKEN_TYPES.RIGHT_PARENTHESIS,
	EXPONENTIATION: LexerTOKEN_TYPES.EXPONENTIATION,
	MULTIPLICATION: LexerTOKEN_TYPES.MULTIPLICATION,
	DOT: LexerTOKEN_TYPES.DOT,
	DIVISION: LexerTOKEN_TYPES.DIVISION,
	REMAINDER: LexerTOKEN_TYPES.REMAINDER,
	ADDITION: LexerTOKEN_TYPES.ADDITION,
	SUBTRACTION: LexerTOKEN_TYPES.SUBTRACTION,
	LESS_THAN: LexerTOKEN_TYPES.LESS_THAN,
	GREATER_THAN: LexerTOKEN_TYPES.GREATER_THAN,
	EQUAL: LexerTOKEN_TYPES.EQUAL,
	NOT_EQUAL: LexerTOKEN_TYPES.NOT_EQUAL,
	LESS_THAN_EQUAL: LexerTOKEN_TYPES.LESS_THAN_EQUAL,
	GREATER_THAN_EQUAL: LexerTOKEN_TYPES.GREATER_THAN_EQUAL,
	LOGICAL_AND: LexerTOKEN_TYPES.LOGICAL_AND,
	LOGICAL_OR: LexerTOKEN_TYPES.LOGICAL_OR,
	LOGICAL_NOT: LexerTOKEN_TYPES.LOGICAL_NOT,
	BITWISE_AND: LexerTOKEN_TYPES.BITWISE_AND,
	BITWISE_OR: LexerTOKEN_TYPES.BITWISE_OR,
	BITWISE_XOR: LexerTOKEN_TYPES.BITWISE_XOR,
	BITWISE_NOT: LexerTOKEN_TYPES.BITWISE_NOT,
	BITWISE_LEFT_SHIFT: LexerTOKEN_TYPES.BITWISE_LEFT_SHIFT,
	BITWISE_RIGHT_SHIFT_ARITHMETIC: LexerTOKEN_TYPES.BITWISE_RIGHT_SHIFT_ARITHMETIC,
	BITWISE_RIGHT_SHIFT_LOGICAL: LexerTOKEN_TYPES.BITWISE_RIGHT_SHIFT_LOGICAL,
	TRUE: LexerTOKEN_TYPES.TRUE,
	FALSE: LexerTOKEN_TYPES.FALSE,
	NUMBER: LexerTOKEN_TYPES.NUMBER,
	IDENTIFIER: LexerTOKEN_TYPES.IDENTIFIER,
	COMMA: LexerTOKEN_TYPES.COMMA,
} as const
export type lexerCalculationTokens = (typeof lexerCalculationTokens)[keyof typeof lexerCalculationTokens]
