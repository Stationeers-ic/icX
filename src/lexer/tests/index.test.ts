import { describe, test, expect } from "bun:test"
import { getNextToken, parse, Token } from "../index"
import RandExp, { randexp } from "randexp"
import { TOKEN_TYPES, TOKENS } from "../tokens"

describe("getNextToken", () => {
	test("Test", () => {
		expect(getNextToken(0, "5")).toMatchSnapshot()
		expect(getNextToken(1, "5i")).toMatchSnapshot()
	})
})
describe("parse", () => {
	test("allStringTokens", () => {
		TOKENS.filter((t) => t.patternType === "string").forEach((token) => {
			try {
				expect(parse(token.pattern)[0]?.[0]?.type).toBe(token.token)
			} catch (e: any) {
				console.log("On test token:", token)
				throw e.message
			}
		})
	})
	describe("allRangeTokens", () => {
		test("basic", () => {
			TOKENS.filter((t) => t.patternType === "range").forEach((token) => {
				expect(parse(`${token.open} ${token.close}`)[0]?.[0]?.type).toBe(token.token)

				const testString = new RandExp(".{5,50}[^\\\\]")
				;(token.open + token.close).split("").forEach((item) => {
					testString.defaultRange.subtract(item.charCodeAt(0))
				})
				for (let i = 0; i < 50; i++) {
					const v = `${token.open}${testString.gen()}${token.close}`
					try {
						expect(parse(v)[0]?.[0]?.type).toBe(token.token)
						expect(parse(v)[0]?.length).toBe(2)
					} catch (e: any) {
						console.log("On test token:", token, "With value:", v)
						throw e.message
					}
				}
			})
		})
		test("ignore", () => {
			TOKENS.filter((t) => t.patternType === "range").forEach((token) => {
				const ignore = token.ignore
				if (!ignore) return
				expect(parse(`${token.open} ${token.close}`)[0]?.[0]?.type).toBe(token.token)

				const testString = new RandExp(".{2,8}[^\\\\]")
				;(token.open + token.close).split("").forEach((item) => {
					testString.defaultRange.subtract(item.charCodeAt(0))
				})
				for (let i = 0; i < 50; i++) {
					ignore.forEach((item) => {
						const v = `${token.open}${item}${testString.gen()}${item}${testString.gen()}${item}${token.close}`
						try {
							expect(parse(v)[0]?.[0]?.type).toBe(token.token)
							expect(parse(v)[0]?.length).toBe(2)
						} catch (e: any) {
							console.log("On test token:", token, "With value:", v)
							throw e.message
						}
					})
				}
			})
		})
		test("errorOn", () => {
			TOKENS.filter((t) => t.patternType === "range").forEach((token) => {
				const errorOn = token.errorOn ?? []
				expect(parse(`${token.open} ${token.close}`)[0]?.[0]?.type).toBe(token.token)

				const testString = new RandExp(".{2,8}[^\\\\]")
				;(token.open + token.close).split("").forEach((item) => {
					testString.defaultRange.subtract(item.charCodeAt(0))
				})
				for (let i = 0; i < 50; i++) {
					;[...errorOn, "\u0003"].forEach((item) => {
						const v = `${token.open}${testString.gen()}${item}${testString.gen()}${token.close}`
						try {
							expect(parse(v)[0]?.[0]?.type).toBe(token.token)
							if (item !== "\u0003") expect(parse(v)[0]?.length).not.toBe(2)
							else expect(parse(v)[0]?.length).toBe(2)
						} catch (e: any) {
							console.log("On test token:", token, "With value:", v)
							item === "\u0003" && console.log("(\\u0003)")
							throw e.message
						}
					})
				}
			})
		})
	})
	describe("allFunctionTokens", () => {
		const testCases: Record<string, [string, len: number, value:any][]> = {}
		testCases[TOKEN_TYPES.COMMENT] = [
			["// comment", 2],
			["//comment\n10", 4],
		]
		testCases[TOKEN_TYPES.NUMBER] = [
			["9.1"]




		]
		const usedCases: string[] = []
		test("TestCases", () => {
			TOKENS.filter((t) => t.patternType === "function").forEach((token) => {
				usedCases.push(token.token)
				try {
					expect(Object.keys(testCases)).toContain(token.token)
				} catch (e: any) {
					console.log("No test cases for:", token.token)
					throw e.message
				}
				testCases[token.token].forEach(([c, len]) => {
					try {
						expect(parse(c)[0]?.[0]?.type).toBe(token.token)
						expect(parse(c)[0]?.length).toBe(len)
					} catch (e: any) {
						console.log("On test case:", c)
						throw e.message
					}
				})
			})
		})

		test("AllCasesUsed", () => {
			usedCases.forEach((key) => {
				try {
					expect(usedCases).toContain(key)
				} catch (e: any) {
					console.log("Unused case:", key)
					throw e.message
				}
			})
		})
	})
	describe("generalTokens", () => {
		;["100", "i100", "i 100", "'100'", '"100"', "x()", "5+x (5-5)", "+x", "x+", "x"].forEach((c) => {
			test(c, () => {
				expect(parse(c)).toMatchSnapshot(c)
			})
		})
		test("empty", () => {
			expect(parse("")).toMatchSnapshot()
		})
	})
})
