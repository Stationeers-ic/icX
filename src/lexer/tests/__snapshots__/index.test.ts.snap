// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getNextToken Test 1`] = `
[
  1,
  {
    "type": "NUMBER",
    "value": 5,
  },
]
`;

exports[`getNextToken Test 2`] = `
[
  1,
  {
    "type": "IDENTIFIER",
    "value": "i",
  },
]
`;

exports[`parse generalTokens 100: 100 1`] = `
[
  [
    {
      "end": 3,
      "length": 3,
      "start": 0,
      "type": "NUMBER",
      "value": 100,
    },
    {
      "end": 3,
      "length": 0,
      "start": 3,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens i100: i100 1`] = `
[
  [
    {
      "end": 4,
      "length": 4,
      "start": 0,
      "type": "IDENTIFIER",
      "value": "i100",
    },
    {
      "end": 4,
      "length": 0,
      "start": 4,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens i 100: i 100 1`] = `
[
  [
    {
      "end": 1,
      "length": 1,
      "start": 0,
      "type": "IDENTIFIER",
      "value": "i",
    },
    {
      "end": 5,
      "length": 3,
      "start": 2,
      "type": "NUMBER",
      "value": 100,
    },
    {
      "end": 5,
      "length": 0,
      "start": 5,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens '100': '100' 1`] = `
[
  [
    {
      "end": 5,
      "length": 5,
      "start": 0,
      "type": "STRING",
      "value": "100",
    },
    {
      "end": 5,
      "length": 0,
      "start": 5,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens "100": "100" 1`] = `
[
  [
    {
      "end": 5,
      "length": 5,
      "start": 0,
      "type": "STRING",
      "value": "100",
    },
    {
      "end": 5,
      "length": 0,
      "start": 5,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens x(): x() 1`] = `
[
  [
    {
      "end": 1,
      "length": 1,
      "start": 0,
      "type": "IDENTIFIER",
      "value": "x",
    },
    {
      "end": 2,
      "length": 1,
      "start": 1,
      "type": "LEFT_PARENTHESIS",
    },
    {
      "end": 3,
      "length": 1,
      "start": 2,
      "type": "RIGHT_PARENTHESIS",
    },
    {
      "end": 3,
      "length": 0,
      "start": 3,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens 5+x (5-5): 5+x (5-5) 1`] = `
[
  [
    {
      "end": 1,
      "length": 1,
      "start": 0,
      "type": "NUMBER",
      "value": 5,
    },
    {
      "end": 2,
      "length": 1,
      "start": 1,
      "type": "ADDITION",
    },
    {
      "end": 3,
      "length": 1,
      "start": 2,
      "type": "IDENTIFIER",
      "value": "x",
    },
    {
      "end": 5,
      "length": 1,
      "start": 4,
      "type": "LEFT_PARENTHESIS",
    },
    {
      "end": 6,
      "length": 1,
      "start": 5,
      "type": "NUMBER",
      "value": 5,
    },
    {
      "end": 7,
      "length": 1,
      "start": 6,
      "type": "SUBTRACTION",
    },
    {
      "end": 8,
      "length": 1,
      "start": 7,
      "type": "NUMBER",
      "value": 5,
    },
    {
      "end": 9,
      "length": 1,
      "start": 8,
      "type": "RIGHT_PARENTHESIS",
    },
    {
      "end": 9,
      "length": 0,
      "start": 9,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens +x: +x 1`] = `
[
  [
    {
      "end": 1,
      "length": 1,
      "start": 0,
      "type": "ADDITION",
    },
    {
      "end": 2,
      "length": 1,
      "start": 1,
      "type": "IDENTIFIER",
      "value": "x",
    },
    {
      "end": 2,
      "length": 0,
      "start": 2,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens x+: x+ 1`] = `
[
  [
    {
      "end": 1,
      "length": 1,
      "start": 0,
      "type": "IDENTIFIER",
      "value": "x",
    },
    {
      "end": 2,
      "length": 1,
      "start": 1,
      "type": "ADDITION",
    },
    {
      "end": 2,
      "length": 0,
      "start": 2,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens x: x 1`] = `
[
  [
    {
      "end": 1,
      "length": 1,
      "start": 0,
      "type": "IDENTIFIER",
      "value": "x",
    },
    {
      "end": 1,
      "length": 0,
      "start": 1,
      "type": "EOF",
    },
  ],
  [],
]
`;

exports[`parse generalTokens empty 1`] = `
[
  [
    {
      "end": 0,
      "length": 0,
      "start": 0,
      "type": "EOF",
    },
  ],
  [],
]
`;
